File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cpp
class Trie {
    vector<Trie*>children;
    bool isEnd;
public:
    
    
    Trie() {
        isEnd=false;
        children= vector<Trie*>(26,NULL);
    }
    
    void insert(string word) {
        Trie *curr= this;
        for(char c:word){
            if(curr->children[c-'a']==NULL)curr->children[c-'a']=new Trie();
            curr= curr->children[c-'a'];
        }
        curr->isEnd=true;
    }
    
    bool search(string word) {
        Trie*curr=this;
        for(char c:word){
            if(curr->children[c-'a']==NULL)return false;
            curr=curr->children[c-'a'];
        }
        return curr && curr->isEnd;
    }
    
    bool startsWith(string prefix) {
        Trie*curr=this;
        for(char c:prefix){
            if(curr->children[c-'a']==NULL)return false;
            curr=curr->children[c-'a'];
        }
        return curr;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.java
class Trie {
    public Trie[] children;
    boolean isEnd;
    public Trie() {
        children=new Trie[26];
        isEnd=false;
    }
    
    public void insert(String word) {
        Trie t=this;
        for(char c:word.toCharArray()){
            if(t.children[c-'a']==null)t.children[c-'a']=new Trie();
            t=t.children[c-'a'];
        }
        t.isEnd=true;
    }
    
    public boolean search(String word) {
        Trie t=this;
        for(char c:word.toCharArray()){
            if(t.children[c-'a']==null)return false;
            t=t.children[c-'a'];
        }
        return t!=null && t.isEnd==true;
    }
    
    public boolean startsWith(String prefix) {
        Trie t=this;
        for(char c:prefix.toCharArray()){
            if(t.children[c-'a']==null)return false;
            t=t.children[c-'a'];
        }
        return t!=null;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */